// This file is automatically generated by sila2lib.proto_builder.proto_builder version 0.2.0
// :generation date: 2019-10-05 17:59:19
// 
// ---- PLEASE DO NOT MODIFY MANUALLY !! ---

syntax = "proto3";
import "SiLAFramework.proto";
package sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1;

// Feature: Pump Fluid Dosing Service
//   
//           Allows to dose a specified fluid. There are commands for absolute dosing (SetFillLevel) and relative dosing
//   (DoseVolume and GenerateFlow) available.
//   
//           The flow rate can be negative. In this case the pump aspirates the fluid instead of dispensing. The flow rate
//   has to be a value between MaxFlowRate and MinFlowRate. If the value is not within this range (hence is invalid) a
//   ValidationError will be thrown.
//           At any time the property CurrentSyringeFillLevel can be queried to see how much fluid is left in the syringe.
//   Similarly the property CurrentFlowRate can be queried to get the current flow rate at which the pump is dosing.
//   
service PumpFluidDosingService {
    // Set Fill Level
    //   
    //                   Pumps fluid with the given flow rate until the requested fill level is reached.
    //                   Depending on the requested fill level given in the FillLevel parameter this function may cause
    //   aspiration or dispension of fluid.
    //   
    rpc SetFillLevel(sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.SetFillLevel_Parameters) returns (sila2.org.silastandard.CommandConfirmation) {}
    rpc SetFillLevel_Info(sila2.org.silastandard.CommandExecutionUUID) returns (stream sila2.org.silastandard.ExecutionInfo) {}
    rpc SetFillLevel_Result(sila2.org.silastandard.CommandExecutionUUID) returns (sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.SetFillLevel_Responses) {}
    // Dose Volume
    //   Dose a certain amount of volume with the given flow rate.
    rpc DoseVolume(sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.DoseVolume_Parameters) returns (sila2.org.silastandard.CommandConfirmation) {}
    rpc DoseVolume_Info(sila2.org.silastandard.CommandExecutionUUID) returns (stream sila2.org.silastandard.ExecutionInfo) {}
    rpc DoseVolume_Result(sila2.org.silastandard.CommandExecutionUUID) returns (sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.DoseVolume_Responses) {}
    // Generate Flow
    //   
    //               Generate a continous flow with the given flow rate. Dosing continues until it gets stopped manually by
    //   calling StopDosage or until the pusher reached one of its limits.
    //   
    rpc GenerateFlow(sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.GenerateFlow_Parameters) returns (sila2.org.silastandard.CommandConfirmation) {}
    rpc GenerateFlow_Info(sila2.org.silastandard.CommandExecutionUUID) returns (stream sila2.org.silastandard.ExecutionInfo) {}
    rpc GenerateFlow_Result(sila2.org.silastandard.CommandExecutionUUID) returns (sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.GenerateFlow_Responses) {}
    // Stop Dosage
    //   Stops a currently running dosage immediately.
    rpc StopDosage(sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.StopDosage_Parameters) returns (sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.StopDosage_Responses) {}
    // Maximum Syringe Fill Level
    //   The maximum amount of fluid that the syringe can hold.
    rpc Subscribe_MaxSyringeFillLevel(sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.Subscribe_MaxSyringeFillLevel_Parameters) returns (stream sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.Subscribe_MaxSyringeFillLevel_Responses) {}
    // Syringe Fill Level
    //   The current amount of fluid left in the syringe.
    rpc Subscribe_SyringeFillLevel(sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.Subscribe_SyringeFillLevel_Parameters) returns (stream sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.Subscribe_SyringeFillLevel_Responses) {}
    // Maximum Flow Rate
    //   The maximum value of the flow rate at which this pump can dose a fluid.
    rpc Subscribe_MaxFlowRate(sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.Subscribe_MaxFlowRate_Parameters) returns (stream sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.Subscribe_MaxFlowRate_Responses) {}
    // Flow Rate
    //   The current value of the flow rate. It is 0 if the pump does not dose any fluid.
    rpc Subscribe_FlowRate(sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.Subscribe_FlowRate_Parameters) returns (stream sila2.de.cetoni.pumps.syringepumps.pumpfluiddosingservice.v1.Subscribe_FlowRate_Responses) {}
    
}

// ----------------- Data Type definitions -----------------


// ------ Command Parameter and Response definitions -------
message SetFillLevel_Parameters {
    sila2.org.silastandard.Real FillLevel = 1;
    sila2.org.silastandard.Real FlowRate = 2;
}

message SetFillLevel_Responses {
    
}

message DoseVolume_Parameters {
    sila2.org.silastandard.Real Volume = 1;
    sila2.org.silastandard.Real FlowRate = 2;
}

message DoseVolume_Responses {
    
}

message GenerateFlow_Parameters {
    sila2.org.silastandard.Real FlowRate = 1;
}

message GenerateFlow_Responses {
    
}

message StopDosage_Parameters {
    
}

message StopDosage_Responses {
    
}
message Subscribe_MaxSyringeFillLevel_Parameters {
    // Empty message
}

message Subscribe_MaxSyringeFillLevel_Responses {
    sila2.org.silastandard.Real MaxSyringeFillLevel = 1;
}

message Subscribe_SyringeFillLevel_Parameters {
    // Empty message
}

message Subscribe_SyringeFillLevel_Responses {
    sila2.org.silastandard.Real SyringeFillLevel = 1;
}

message Subscribe_MaxFlowRate_Parameters {
    // Empty message
}

message Subscribe_MaxFlowRate_Responses {
    sila2.org.silastandard.Real MaxFlowRate = 1;
}

message Subscribe_FlowRate_Parameters {
    // Empty message
}

message Subscribe_FlowRate_Responses {
    sila2.org.silastandard.Real FlowRate = 1;
}

// ----------------- Metadata Definitions ------------------
